src/test/python/test_signer.py

```diff
@@ -0,0 +1,183 @@
+
+
+def test_malformed_payload():
+    with pytest.raises(ValueError) as excepinfo:
+        signer.__get_hashed_payload(b'{""}')
+        assert excepinfo.type is json.JSONDecodeError
+
+
+def test_none_payload():
+    assert signer.__get_hashed_payload(b'{}') == \
+           signer.__get_hashed_payload(None)
+
+
+def test_unordered_payload():
+    assert signer.__get_hashed_payload(b'{"name": "rattan", "a": "value"}') == \
+           signer.__get_hashed_payload(b'{"a": "value", "name": "rattan"}')
+
+
+def test_nested_payload():
+    input_json = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    sorted_json = \
+        b"""
+        {
+            "a": "value",
+            "c": {
+                "a": "hello",
+                "za": "value"
+            },
+            "name": "rattan"
+        }
+        """
+
+    assert signer.__get_hashed_payload(input_json) == \
+           signer.__get_hashed_payload(sorted_json)
+
+
+@pytest.mark.parametrize('url, expected_encoded_url', [
+    ('/identity blah/', '/identity%20blah/'),
+    ('/identity-blah/', '/identity-blah/'),
+    ('/', '/'),
+    ('/identityABC/', '/identityABC/'),
+])
+def test_uri_endoing(url, expected_encoded_url):
+    assert signer.__encode_uri(url) == expected_encoded_url
+
+
+def test_signature_material(private_key):
+    method = "POST"
+    url = "https://delta.covata.io/v1/secrets?hello=w+orld"
+    headers = {
+        "Content-Type": "application/json"
+    }
+    payload = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    with freeze_time():
+        url_parsed = urllib.parse.urlparse(url)
+        cvt_date = datetime.utcnow().strftime(CVT_DATE_FORMAT)
+        headers_ = dict(headers)
+        headers_["Cvt-Date"] = cvt_date
+        uri = signer.__encode_uri("/".join(url_parsed.path.split("/")[2:]))
+        query = url_parsed.query.replace("+", "%20")
+        canonical_headers = 'content-type:application/json\n cvt-date:' \
+                            + cvt_date
+        signed_headers = 'content-type;cvt-date'
+        hashed_payload = \
+            '758ffa295b9a475f04aef51abd60563dabb8df1988cf6d62b9298b1d5ba6b8bf'
+        materials = signer.SignatureMaterial(
+            method=method,
+            uri=uri,
+            headers_=headers_,
+            query_params=query,
+            canonical_headers=canonical_headers,
+            signed_headers=signed_headers,
+            hashed_payload=hashed_payload,
+            cvt_date=cvt_date)
+        materials_from_signer = \
+            signer.__get_signature_materials(method, url, headers, payload)
+        assert materials.canonical_request == \
```

Comment:
I'll raise this as a refactor issue.

I think we can still do a header replacement of the whole dictionary, all it means is the caller constructs the dictionary (rather than passing in r.headers). I am ok with shifting the responsibility of ensuring cvt-date and host onto the caller's responsibility and absolve the signing module of time-dependent calls within the methods.